File: fb/22.Generate-Parentheses.cpp
// Backtracking hard to estimate complexity. Details see leetcode solution page
class Solution {
private:
    void helper(int l, int r, string& cur, vector<string>& ans) {
        if(l == 0 && r == 0) {
            ans.push_back(cur);
            return;
        }
        int len = cur.size();
        if(l > 0) {
            cur += '(';
            helper(l - 1, r, cur, ans);
            cur.resize(len);
        }
        if(l < r) {
            cur += ')';
            helper(l, r - 1, cur, ans);
            cur.resize(len);
        }
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string cur = "";
        helper(n, n, cur, ans);
        return ans;
    }
};

File: fb/22.Generate-Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        def backtracking(s=[], left=0, right=0):
            if len(s) == n*2:
                ans.append("".join(s))
                return
            if left < n:
                s.append("(")
                backtracking(s, left+1, right)
                s.pop()
            if right < left:
                s.append(")")
                backtracking(s, left, right+1)
                s.pop()
        backtracking()
        return ans
